src/doc/man/recollq.1
.\" $Id: recollq.1,v 1.1 2007-11-13 10:07:35 dockes Exp $ (C) 2005 J.F.Dockes\$
.TH RECOLLQ 1 "13 November 2007"
.SH NAME
recollq \- command line / standard output Recoll query command.
.SH SYNOPSIS
.B recollq
[
.B \-c
<configdir>
]
[
.B \-o
|
.B \-f
|
.B \-a
]
[
.B \-b
]
[
.B \-d
]
[
.B \-A
]
[
.B \-p
]
[
.B \-e
]
[
.B \-m
]
[
.B \-n
<[first-]cnt>
]
[
.B \-Q
]
[
.B \-s
<stemming language>
]
[
.B \-S
<fldname>
]
[
.B \-D
]
[
.B \-i
<additional index directory>
]
[
.B \-F
<quoted space separated field name list>
]
[
.B \--extract-to
<file path>
]
<query string>

.B recollq \-P

.SH DESCRIPTION
The
.B recollq
command will execute the Recoll query specified on the command line and
print the results to the standard output. It is primarily designed for
diagnostics, or piping the data to some other program. The basic format and
its variations can be useful for command line querying. The \-F option
should exclusively be used for using the output data in another program, as
it is the only one for which output is guaranteed to be fully parseable.
.PP
The 
.B \-c 
option specifies the configuration directory name, overriding the
default or $RECOLL_CONFDIR.
.PP
The query string is built by concatenating all arguments found at the end
of the command line (after the options). It will be interpreted by default
as a query language string. Quoting should be used as needed to escape
characters that might be interpreted by the shell (ie: wildcards).
.B \-a 
is specified, the query string will be interpreted as an
.I all words
simple search query. If 
.B \-o 
is specified, the query string will be interpreted as an
.I any word
simple search query. If 
.B \-f
is specified, the query string will be interpreted as a
.I file name
simple search query. 
.PP
.B \-b
(basic) can be specified to only print the result urls in the output
stream.
.PP
If 
.B \-d
is set, the text for the result files contents will be dumped to stdout.
.PP
If 
.B \-m
is set, the whole metadata array will be dumped for each document.
.PP
If 
.B \-A
is set, the document abstracts will be printed. With an additional
.B \-p
option, snippets with page numbers (when available) will be shown instead.
.PP
.B \-S
<fieldname>
sorts the results according to the specified field. Use 
.B \-D 
for descending order.
.PP
.B \-n
<[first-]cnt>
can be used to set the maximum number of results that should be
printed. The default is 2000. Use a value of 0 for no limit. If the argument is of the form
first-cnt, it also defines the first result to output (from 0).
.PP
.B \-s
<language>
selects the word stemming language. The value should match an existing
stemming database (as set in the configuration or added with recollindex \-s).
.PP
.B \-i
<extra dbdir>
adds the specified Xapian index to the set used for the query. Can be
specified multiple times.
.PP 
.B \-F
<space separated field list> (one argument, e.g. "author title")
should be used for piping the data to another program. After 2 initial
lines showing the actual query and the estimated result counts, it will
print one line for each result document. Each line will have 
exactly the fields requested on the command line. Fields are encoded in
base64 and separated by one space character. Empty fields are indicated by
consecutive space characters. There is one additional space character at
the end of each line.
.PP
.B \--extract-to
<file path>
Will extract the first result document of the query to the argument path, which must not exist. Use
-n first-cnt to select the document.
.PP
.B recollq \-P
(Period) will print the minimum and maximum modification years for
documents in the index.

.SH SEE ALSO
.PP 
recollindex(1) recollq(1) recoll.conf(5) 


src/ChangeLog
2009-10-19 12:00  dockes

	* ChangeLog: release 1.12.2 (pre?)

2009-10-19 10:51  dockes

	* qtgui/uiprefs.ui: move the use-desktop-preference checkbox close
	  to the choose editors button

2009-10-19 07:25  dockes

	* rcldb/rcldb.cpp, rcldb/rclquery.cpp: catch xapian exceptions in 2
	  more places.

2009-10-18 07:57  dockes

	* qtgui/reslist.cpp: reslist: rightclick popup would not work
	  inside table

2009-10-17 06:35  dockes

	* Makefile.in, index/recollindex.cpp, qtgui/main.cpp,
	  qtgui/rclmain_w.cpp, rcldb/rcldb.cpp, rcldb/rcldb.h: rclversion.h
	  must not include xapian.h. Replace with Rcl::version_string()

2009-10-17 06:34  dockes

	* ChangeLog:

2009-10-15 12:28  dockes

	* index/rclmonprc.cpp, utils/Makefile, utils/conftree.cpp,
	  utils/execmd.cpp: only call x11IsAlive() from main thread

2009-10-14 14:41  dockes

	* sampleconf/recoll.conf.in:

2009-10-14 14:41  dockes

	* internfile/mh_exec.cpp: no timeout if filtermaxseconds<0

2009-10-13 12:17  dockes

	* internfile/mh_exec.cpp: typo in exec timeout

2009-10-13 11:52  dockes

	* sampleconf/recoll.conf.in: add loop.ps to skippedNames

2009-10-12 17:41  dockes

	* internfile/mh_exec.cpp: configurable timeout on script execution

2009-10-12 16:52  dockes

	* index/recollindex.cpp, internfile/mh_exec.cpp, utils/execmd.cpp:
	  Handle the case of runaway filter processes (ie: loop.ps):
	  - Modify ExecCmd to use a different process group for the son
	  process
	  and use killpg for termination (ensure that all children get
	  killed,
	  not just the top one)
	  - Modify ExecCmd to test for term request on a select() error
	  (avoid calling
	  waitpid() without prior killing in this case).
	  - Modify mh_exec to checkCancel() in the heartbeat/data arrival
	  routine
	  - Modify recollindex to call setCancel when catching a signal

2009-10-12 16:26  dockes

	* common/rclinit.cpp: also block USR1 USR2

2009-10-12 16:26  dockes

	* VERSION:

2009-10-12 11:29  dockes

	* common/textsplit.cpp: merged test driver from 1.13 (no change to
	  main code

2009-08-13 06:36  dockes

	* index/indexer.cpp, internfile/internfile.cpp,
	  internfile/internfile.h, internfile/mimehandler.cpp,
	  internfile/mimehandler.h: xattrs: make them work with non-text
	  files. Use ctime for up to date checks

2009-08-13 06:21  dockes

	* qtgui/rclmain_w.cpp: Make sure db is open at all times (caused
	  problems when sorting query started from the command line)

2009-08-13 06:19  dockes

	* sampleconf/recoll.conf.in: add indexedmimetypes to sample file

2009-07-02 06:18  dockes

	* qtgui/main.cpp, rcldb/searchdata.h, utils/mimeparse.cpp,
	  utils/mimeparse.h: explicitely->explicitly

2009-06-26 09:25  dockes

	* filters/rclman, sampleconf/mimemap: improve man page handling

2009-06-23 06:00  dockes

	* README:

2009-06-23 05:52  dockes

	* ChangeLog, VERSION: version 1.12.1

2009-06-22 08:59  dockes

	* filters/rclman: use groff html output!

2009-06-01 06:32  dockes

	* qtgui/main.cpp: fixed bug in handling remaining arguments as
	  question pieces

2009-05-29 06:28  dockes

	* bincimapmime/convert.h, utils/base64.cpp: change strchr() return
	  parameter to const for new libc

2009-05-25 09:00  dockes

	* filters/rcllyx: bug report from d.prost: spaces and accents in
	  lyx file names

2009-05-04 08:06  dockes

	* qtgui/guiutils.cpp, qtgui/preview_w.h, qtgui/spell_w.cpp: gcc44

2009-04-27 10:42  dockes

	* utils/refcntr.h: MFC

2009-04-27 10:39  dockes

	* utils/md5.cpp: MFC

2009-04-27 10:38  dockes

	* utils/readfile.cpp: MFC

2009-04-27 10:38  dockes

	* utils/refcntr.h: MFC

2009-04-27 10:06  dockes

	* qtgui/i18n, qtgui/i18n/recoll_ru.ts, qtgui/i18n/recoll_uk.ts:
	  brought trunk changes into 1.12maint

2009-04-27 09:23  dockes

	* python/recoll/setup.py: pathhash->fileudi

2009-04-27 09:20  dockes

	* Maintenance branch for 1.12

2009-01-30 13:28  dockes

	* Release 1.12.0 tagged

2009-01-30 11:42  dockes

	* 1.12 manual

2009-01-30 10:22  dockes

	* updated message files, translated french

2009-01-29 14:24  dockes

	* simplified rcltext. No need for awk and no assumptions on charset

2009-01-29 11:27  dockes

	* 

2009-01-29 11:04  dockes

	* 1.12.0 une

2009-01-29 10:08  dockes

	* have ssearch install the lang help section when needed

2009-01-28 17:41  dockes

	* erase history would crash with empty reslist docsource

2009-01-28 17:21  dockes

	* fixed status bar messages (were cleared by periodic100 every
	  100ms)

2009-01-28 17:05  dockes

	* F1 context-enhanced help

2009-01-28 16:56  dockes

	* F1 context-enhanced help

2009-01-28 16:56  dockes

	* F1 context-enhanced help

2009-01-28 14:58  dockes

	* dont encode urls used for starting help browser

2009-01-28 14:22  dockes

	* add xapian version to version string

2009-01-28 08:45  dockes

	* avoid setting 0 sizes

2009-01-28 08:40  dockes

	* allow setting QMAKE in the environment

2009-01-27 18:12  dockes

	* 

2009-01-27 10:25  dockes

	* Emit a_b intermediary span when splitting a_b.c

2009-01-26 18:30  dockes

	* modified the time at which we unaccent so that we can do the
	  Capitalized->nostemming test on single words (this had been
	  broken by the change of noac/split order done earlier to get
	  japanese to work)

2009-01-26 13:29  dockes

	* add overloaded neutchars with different parameters

2009-01-26 13:27  dockes

	* tested and decided against cacheing iconv_open

2009-01-23 15:56  dockes

	* temp ckpt: need to test on real unix

2009-01-23 11:07  dockes

	* avoid name duplication

2009-01-23 11:03  dockes

	* one button for choosing native editors

2009-01-23 10:38  dockes

	* simplified javascrip: no ie here!

2009-01-23 09:41  dockes

	* toLocal8Bit->local8bit

2009-01-23 09:30  dockes

	* use normal text/html ext app for viewing help

2009-01-23 09:27  dockes

	* accept additional path argument to execmd::which

2009-01-22 14:25  dockes

	* allow toggle show text/fields in preview

2009-01-21 16:42  dockes

	* added saveToFile menu entry to reslist


src/query/recollq.cpp
/* Copyright (C) 2006-2022 J.F.Dockes
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the
 *   Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
// Takes a query and run it, no gui, results to stdout
#include "autoconfig.h"

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <limits.h>
#include <getopt.h>

#include <iostream>
#include <sstream>
#include <list>
#include <string>

#include "rcldb.h"
#include "rclquery.h"
#include "rclconfig.h"
#include "pathut.h"
#include "rclinit.h"
#include "log.h"
#include "wasatorcl.h"
#include "internfile.h"
#include "wipedir.h"
#include "transcode.h"
#include "textsplit.h"
#include "smallut.h"
#include "chrono.h"
#include "base64.h"
#include "rclutil.h"
#include "internfile.h"

using namespace std;

bool dump_contents(RclConfig *rclconfig, Rcl::Doc& idoc)
{
    FileInterner interner(idoc, rclconfig, FileInterner::FIF_forPreview);
    Rcl::Doc fdoc;
    string ipath = idoc.ipath;
    if (interner.internfile(fdoc, ipath)) {
        cout << fdoc.text << "\n";
    } else {
        cout << "Cant turn to text:" << idoc.url << " | " << idoc.ipath << "\n";
    }
    return true;
}

string make_abstract(Rcl::Doc& doc, Rcl::Query& query, bool asSnippets,
                     int snipcount, bool showlines)
{
    string abstract;
    if (asSnippets) {
        std::vector<Rcl::Snippet> snippets;
        std::ostringstream str;
        if (query.makeDocAbstract(doc, snippets, snipcount, -1, true)) {
            for (const auto& snippet : snippets) {
                str << (showlines ? snippet.line : snippet.page) << " : " << snippet.snippet << "\n";
            }
        }
        abstract = str.str();
    } else {
        query.makeDocAbstract(doc, abstract);
        abstract += "\n";
    }
    return abstract;
}

void output_fields(vector<string> fields, Rcl::Doc& doc,
                   Rcl::Query& query, Rcl::Db&, bool printnames,
                   bool asSnippets, int snipcnt, bool showlines)
{
    if (fields.empty()) {
        for (const auto& entry : doc.meta) {
            fields.push_back(entry.first);
        }
    }
    for (const auto& fld : fields) {
        string out;
        if (fld == "abstract") {
            base64_encode(make_abstract(doc, query, asSnippets, snipcnt, showlines), out);
        } else if (fld == "xdocid") {
            char cdocid[30];
            sprintf(cdocid, "%lu", (unsigned long)doc.xdocid);
            base64_encode(cdocid, out);
        } else {
            base64_encode(doc.meta[fld], out);
        }
        // Before printnames existed, recollq printed a single blank for empty
        // fields. This is a problem when printing names and using strtok, but
        // have to keep the old behaviour when printnames is not set.
        if (!(out.empty() && printnames)) {
            if (printnames) {
                cout << fld << " ";
            }
            cout << out << " ";
        }
    }
    cout << "\n";
}

static char *thisprog;
static char usage [] =
" -P: Show the date span for all the documents present in the index.\n"
" [-o|-a|-f] [-q] <query string>\n"
" Runs a recoll query and displays result lines. \n"
"  Default: will interpret the argument(s) as a xesam query string.\n"
"  Query elements: \n"
"   * Implicit AND, exclusion, field spec:  t1 -t2 title:t3\n"
"   * OR has priority: t1 OR t2 t3 OR t4 means (t1 OR t2) AND (t3 OR t4)\n"
"   * Phrase: \"t1 t2\" (needs additional quoting on cmd line)\n"
" -o Emulate the GUI simple search in ANY TERM mode.\n"
" -a Emulate the GUI simple search in ALL TERMS mode.\n"
" -f Emulate the GUI simple search in filename mode.\n"
" -q is just ignored (compatibility with the recoll GUI command line).\n"
"Common options:\n"
" -c <configdir> : specify config directory, overriding $RECOLL_CONFDIR.\n"
" -C : collapse duplicates\n"            
" -d also dump file contents.\n"
" -n [first-]<cnt> define the result slice. The default value for [first]\n"
"    is 0. Without the option, the default max count is 2000.\n"
"    Use n=0 for no limit.\n"
" -b : basic. Just output urls, no mime types or titles.\n"
" -Q : no result lines, just the processed query and result count.\n"
" -m : dump the whole document meta[] array for each result.\n"
" -A : output the document abstracts.\n"
"    -p <cnt> : show <cnt> snippets, with page numbers instead of the concatenated abstract.\n"
"    -g <cnt> : show <cnt> snippets, with line numbers instead of the concatenated abstract.\n"
" -S fld : sort by field <fld>.\n"
"   -D : sort descending.\n"
" -s stemlang : set stemming language to use (must exist in index...).\n"
"    Use -s \"\" to turn off stem expansion.\n"
" -T <synonyms file>: use the parameter (Thesaurus) for word expansion.\n"
" -i <dbdir> : additional index, several can be given.\n"
" -e use url encoding (%xx) for urls.\n"
" -E use exact result count instead of lower bound estimate.\n"
" -F <field name list> : output exactly these fields for each result.\n"
"    The field values are encoded in base64, output in one line and \n"
"    separated by one space character. This is the recommended format \n"
"    for use by other programs. Use a normal query with option -m to \n"
"    see the field names. Use -F '' to output all fields, but you probably\n"
"    also want option -N in this case.\n"
"   -N : with -F, print the (plain text) field names before the field values.\n"
" --extract_to <filepath> : extract the first result to filepath, which must not exist.\n"
"      Use a -n option with an offset to select the appropriate result.\n"
;

static void Usage(std::ostream &os = std::cerr)
{
    os << thisprog <<  ": usage:" << "\n" << usage;
    exit(1);
}

// BEWARE COMPATIBILITY WITH recoll OPTIONS letters
static int     op_flags;

#define OPT_A 0x1   
#define OPT_a 0x2   
#define OPT_b 0x4   
#define OPT_C 0x8   
#define OPT_D 0x10  
#define OPT_d 0x20  
#define OPT_e 0x40  
#define OPT_F 0x80  
#define OPT_g 0x100 
#define OPT_f 0x200 
#define OPT_l 0x400 
#define OPT_m 0x800 
#define OPT_N 0x1000
#define OPT_o 0x2000
#define OPT_p 0x4000
#define OPT_P 0x8000
#define OPT_Q 0x10000
#define OPT_q 0x20000
#define OPT_S 0x40000
#define OPT_E 0x80000

static struct option long_options[] = {
    #define OPTION_EXTRACT 1000
    {"extract-to", required_argument, 0, OPTION_EXTRACT},
    {0, 0, 0, 0}
};

int recollq(RclConfig **cfp, int argc, char **argv)
{
    thisprog = argv[0];

    string a_config;
    string sortfield;
    string stemlang("english");
    list<string> extra_dbs;
    string sf;
    string syngroupsfn;
    int firstres = 0;
    int maxcount = 2000;
    int snipcnt = -1;
    std::string extractfile;
    
    int ret;
    while ((ret = getopt_long(argc, argv, "+AabCc:DdEefF:hi:lmNn:oPp:g:QqS:s:tT:v",
                              long_options, NULL)) != -1) {
        switch (ret) {
        case 'A': op_flags |= OPT_A; break;
            // GUI: -a same
        case 'a': op_flags |= OPT_a; break;
        case 'b': op_flags |= OPT_b; break;
        case 'C': op_flags |= OPT_C; break;
            // GUI: -c same
        case 'c': a_config = optarg; break;
        case 'd': op_flags |= OPT_d; break;
        case 'D': op_flags |= OPT_D; break;
        case 'E': op_flags |= OPT_E; break;
        case 'e': op_flags |= OPT_e; break;
        case 'F': op_flags |= OPT_F; sf = optarg; break;
            // GUI: -f same
        case 'f': op_flags |= OPT_f; break;
            // GUI: -h same
        case 'h': Usage(std::cout); // Usage exits
        case 'i': extra_dbs.push_back(optarg); break;
            // GUI uses -L to set language of messages
            // GUI: -l specifies query language, which is the default. Accept and ignore
        case 'l': op_flags |= OPT_l; break;
        case 'm': op_flags |= OPT_m; break;
        case 'N': op_flags |= OPT_N; break;
        case 'n':   
        {
            string rescnt = optarg;
            string::size_type dash = rescnt.find("-");
            if (dash != string::npos) {
                firstres = atoi(rescnt.substr(0, dash).c_str());
                if (dash < rescnt.size()-1) {
                    maxcount = atoi(rescnt.substr(dash+1).c_str());
                }
            } else {
                maxcount = atoi(rescnt.c_str());
            }
            if (maxcount <= 0) maxcount = INT_MAX;
        }
        break;
        // GUI: -o same
        case 'o':   op_flags |= OPT_o; break;
        case 'P':   op_flags |= OPT_P; break;
        case 'p':
            op_flags |= OPT_p;
            goto porg;
        case 'g':
            op_flags |= OPT_g;
            {
            porg:
                const char *cp = optarg;
                char *cpe;
                snipcnt = strtol(cp, &cpe, 10);
                if (*cpe != 0)
                    Usage();
            }
            break;
        case 'Q':   op_flags |= OPT_Q; break;
            // GUI: -q same
        case 'q':   op_flags |= OPT_q; break;
        case 'S':   op_flags |= OPT_S; sortfield = optarg; break;
        case 's':   stemlang = optarg; break;
            // GUI: -t use command line, us: ignored
        case 't': break;
        case 'T':   syngroupsfn = optarg; break;
            // GUI: -v same
        case 'v': std::cout << Rcl::version_string() << "\n"; return 0;
            // GUI uses -w : open minimized
        case OPTION_EXTRACT: extractfile=optarg;break;
        }
    }

    string reason;
    *cfp = recollinit(0, 0, 0, reason, &a_config);
    RclConfig *rclconfig = *cfp;
    if (!rclconfig || !rclconfig->ok()) {
        std::cerr << "Recoll init failed: " << reason << "\n";
        exit(1);
    }

    if (argc < 1 && !(op_flags & OPT_P)) {
        Usage();
    }
    vector<string> fields;
    if (op_flags & OPT_F) {
        if (op_flags & (OPT_b|OPT_d|OPT_b|OPT_Q|OPT_m|OPT_A))
            Usage();
        stringToStrings(sf, fields);
    }
    Rcl::Db rcldb(rclconfig);
    if (!extra_dbs.empty()) {
        for (const auto& db : extra_dbs) {
            if (!rcldb.addQueryDb(db)) {
                cerr << "Can't add index: " << db << "\n";
                exit(1);
            }
        }
    }
    if (!syngroupsfn.empty()) {
        if (!rcldb.setSynGroupsFile(syngroupsfn)) {
            cerr << "Can't use synonyms file: " << syngroupsfn << "\n";
            exit(1);
        }
    }
    
    if (!rcldb.open(Rcl::Db::DbRO)) {
        cerr << "Cant open database in " << rclconfig->getDbDir() << 
            " reason: " << rcldb.getReason() << "\n";
        exit(1);
    }

    if (op_flags & OPT_P) {
        int minyear, maxyear;
        if (!rcldb.maxYearSpan(&minyear, &maxyear)) {
            cerr << "maxYearSpan failed: " << rcldb.getReason() << "\n";
            return 1;
        } else {
            cout << "Min year " << minyear << " Max year " << maxyear << "\n";
            return 0;
        }
    }

    if (optind >= argc) {
        Usage();
    }
    string qs;
    while (optind < argc) {
        qs += std::string(argv[optind++]) + " ";
    }

    {
        string uq;
        string charset = rclconfig->getDefCharset(true);
        int ercnt;
        if (!transcode(qs, uq, charset, "UTF-8", &ercnt)) {
            std::cerr << "Can't convert command line args to utf-8\n";
            return 1;
        } else if (ercnt) {
            std::cerr <<
                ercnt << " errors while converting arguments from " << charset << "to utf-8\n";
        }
        qs = uq;
    }

    std::shared_ptr<Rcl::SearchData> sd;

    if (op_flags & (OPT_a|OPT_o|OPT_f)) {
        sd = std::make_shared<Rcl::SearchData>(Rcl::SCLT_OR, stemlang);
        Rcl::SearchDataClause *clp;
        if (op_flags & OPT_f) {
            clp = new Rcl::SearchDataClauseFilename(qs);
        } else {
            clp = new Rcl::SearchDataClauseSimple(
                (op_flags & OPT_o) ? Rcl::SCLT_OR : Rcl::SCLT_AND, qs);
        }
        if (sd)
            sd->addClause(clp);
    } else {
        sd = wasaStringToRcl(rclconfig, stemlang, qs, reason);
    }

    if (!sd) {
        std::cerr << "Query string interpretation failed: " << reason << "\n";
        return 1;
    }

    std::shared_ptr<Rcl::SearchData> rq(sd);
    Rcl::Query query(&rcldb);
    if (op_flags & OPT_C) {
        query.setCollapseDuplicates(true);
    }
    if (op_flags & OPT_S) {
        query.setSortBy(sortfield, (op_flags & OPT_D) ? false : true);
    }
    Chrono chron;
    if (!query.setQuery(rq)) {
        std::cerr << "Query setup failed: " << query.getReason() << "\n";
        return 1;
    }
    int cnt;
    if (op_flags & OPT_E) {
        cnt = query.getResCnt(-1, true);
    } else {
        cnt = query.getResCnt();
    }
    if (!(op_flags & OPT_b)) {
        std::cout << "Recoll query: " << rq->getDescription() << "\n";
        if (firstres == 0) {
            if (cnt <= maxcount)
                std::cout << cnt << " results" << "\n";
            else
                std::cout << cnt << " results (printing  " << maxcount << " max):" << "\n";
        } else {
            std::cout << "Printing at most " << cnt - (firstres+maxcount) <<
                " results from first " << firstres << "\n";
        }
    }
    if (op_flags & OPT_Q) {
        std::cout << "Query setup took " << chron.millis() << " mS" << "\n";
        return 0;
    }

    for (int i = firstres; i < firstres + maxcount; i++) {
        Rcl::Doc doc;
        if (!query.getDoc(i, doc))
            break;

        if (!extractfile.empty()) {
            if (path_exists(extractfile)) {
                std::cerr << "Output file must not exist.\n";
                return 1;
            }
            TempFile unused;
            auto ret = FileInterner::idocToFile(unused, extractfile, rclconfig, doc, false);
            return ret ? 0 : 1;
        }
        
        if (op_flags & OPT_F) {
            output_fields(fields, doc, query, rcldb,
                          op_flags & OPT_N, op_flags & (OPT_p|OPT_g), snipcnt, op_flags & OPT_g);
            continue;
        }

        if (op_flags & OPT_e) 
            doc.url = url_encode(doc.url);

        if (op_flags & OPT_b) {
            cout << doc.url << "\n";
        } else {
            string titleorfn = doc.meta[Rcl::Doc::keytt];
            if (titleorfn.empty())
                titleorfn = doc.meta[Rcl::Doc::keyfn];
            if (titleorfn.empty()) {
                string url;
                printableUrl(rclconfig->getDefCharset(), doc.url, url);
                titleorfn = path_getsimple(url);
            }

            char cpc[20];
            sprintf(cpc, "%d", doc.pc);
            cout 
                << doc.mimetype << "\t"
                << "[" << doc.url << "]" << "\t" 
                << "[" << titleorfn << "]" << "\t"
                << doc.fbytes << "\tbytes" << "\t"
                <<  "\n";
            if (op_flags & OPT_m) {
                for (const auto& ent : doc.meta) {
                    cout << ent.first << " = " << ent.second << "\n";
                }
            }
            if (op_flags & OPT_A) {
                bool asSnippets = (op_flags & (OPT_p|OPT_g)) != 0;
                bool showlines = (op_flags & OPT_g) != 0;
                string abstract = make_abstract(doc, query, asSnippets, snipcnt, showlines);
                string marker = asSnippets ? "SNIPPETS" : "ABSTRACT";
                if (!abstract.empty()) {
                    cout << marker << "\n";
                    cout << abstract;
                    cout << string("/") + marker << "\n";
                }
            }
        }
        if (op_flags & OPT_d) {
            dump_contents(rclconfig, doc);
        }   
    }

    return 0;
}
